### ISSUE TITLE
v0.8.65 EPIC: Fleet execution substrate for profiled workers

### ISSUE THREAD

**Issue by Hmbown** [Hmbown/CodeWhale#3154]
## Goal

Make Fleet the durable execution substrate for profiled CodeWhale workers.

This issue should not be read as "build a separate agent engine" or "own all role/model policy." Fleet profile, role, slot, loadout, permission, and delegation policy live in #3167. Semantic model routing and Fleet model classes live in #3205 and must resolve through #3384. This issue owns the worker lifecycle, ledger, scheduling, receipts, artifacts, heartbeats, retries, and supervision mechanics underneath those policies.

## Architecture Contract

Use one product term: Fleet.

Required layering:

- Fleet profile layer (#3167): roles, slots, personas, tool profiles, permissions, budgets, delegation, escalation, and model/reasoning loadout policy.
- Route layer (#2608, #3205, #3384): provider/model/offering catalogs, semantic route roles, Fleet model classes, provider-scoped route resolution, and `ReadyRouteCandidate` validation.
- Fleet execution layer (this issue): durable worker lifecycle, claims, heartbeats, leases, retries, receipts, artifacts, cancellation, checkpointing, and synthesis hooks.
- WhaleFlow IR layer (#2981, #3229): typed branch/leaf workflow plans that can compile into Fleet tasks.

Do not add:

- A second agent runtime.
- A second progress store parallel to existing subagent/fleet ledgers.
- Fleet-specific provider/model tables.
- Prompt-embedded model routing policy.
- Another naming/API layer for the same worker/profile concept.
- Ad hoc worker state that cannot be replayed or inspected.

## v0.8.65 Scope

The current milestone should make the substrate line up horizontally with the new Fleet profile and routing design:

1. Ensure `FleetTaskSpec` and worker specs can carry a Fleet role plus optional Fleet model class, semantic route role, or explicit provider/model override.
2. Ensure worker launch calls the same route resolver used by TUI/exec/subagents.
3. Persist resolved route details in the ledger: provider, canonical model, wire model, reasoning, route source, Fleet role, Fleet slot, and capability summary.
4. Keep worker authority inherited/narrowed from the parent/session policy; do not inherit model/reasoning blindly.
5. Use the existing Fleet ledger for lifecycle and receipts instead of inventing another store.
6. Add parity tests proving an equivalent role/loadout resolves the same through subagent and Fleet paths.
7. Update docs/runbooks so agents know #3167 is Fleet profile policy, #3205 is routing/classes, and #3154 is execution.

## Execution Responsibilities

Fleet execution owns:

- worker lifecycle: queued, starting, running, model_wait, running_tool, artifact, completed, failed, cancelled, interrupted, stale;
- durable claims, leases, heartbeats, retry budgets, and stale-worker detection;
- scheduler/backpressure over host capacity, provider route, write-capable workers, wall-clock, tokens, and cost budgets;
- artifact and receipt recording;
- cancellation and interruption lineage;
- restart/resume from persisted ledger state;
- planner wakeups on barriers, stale thresholds, verifier results, budget boundaries, and human approvals;
- synthesis/reduce hooks that consume worker receipts without requiring raw transcript replay.

## Product Shape

A user should be able to start a Fleet run, inspect it, stop it, and resume it:

```sh
codewhale fleet run tasks/release-triage.toml --max-workers 8
codewhale fleet status
codewhale fleet inspect <worker-id>
codewhale fleet stop --all
```

The TUI/CLI should show workers grouped by Fleet role/slot and should expose the resolved route source without requiring users or agents to infer provider/model state from naked strings.

## Non-Goals

- No default-on cloud agent farm.
- No secrets in prompts, logs, argv, links, or persisted event payloads.
- No automatic Slack/PagerDuty/webhook writes without explicit config.
- No model-weight training loop in CodeWhale.
- No bypass around Rust-owned sandboxing, permissions, provider calls, cancellation, or event accounting.
- No Fleet-specific provider/model selector.

## Day-Scale Autonomy Contract

Long-running Fleet runs should remain safe and inspectable:

- The manager can resume from durable ledger state after process exit, terminal detach, laptop sleep, network loss, or Runtime API restart.
- Long runs split into checkpoint epochs with bounded context growth: receipts, summaries, artifacts, next actions, and blockers are written before continuing.
- Workers use leases, heartbeats, retry budgets, exponential backoff, and circuit breakers so a stuck provider, dead host, or failing task cannot exhaust the run.
- Cost governors use provider/offering-scoped pricing data with provenance; do not assume any provider is cheap by default.
- Operator escalation is explicit: pause, request approval, or notify a configured channel; never ask for secrets or bypass auth to keep going.
- A run has a stable handoff artifact explaining current scopes, completed receipts, active workers, next leases, and blockers.

## Acceptance Criteria

- A maintainer can run a local multi-worker smoke and see durable status, heartbeats, artifacts, receipts, and terminal outcomes.
- Worker launch uses FleetProfile role/slot/loadout policy plus the shared route resolver.
- Resolved route details are recorded in the ledger for every worker.
- A stuck worker is detected and either restarted, stopped, or escalated according to policy.
- Fleet state is visible from CLI/TUI/API and is replayable enough for handoff/debugging.
- Docs explain how Fleet profiles relate to subagents, Runtime API, and WhaleFlow.

## Verification

Required tests:

- Fleet worker route parity with subagent route resolution.
- Ledger schema/receipt test for resolved provider/model/wire-model/reasoning/source/class/slot.
- Authority narrowing test: tools/permissions/secrets narrow from parent to child, while model/reasoning are resolved by loadout policy.
- Stale worker lease/heartbeat test.
- Resume/replay test from persisted ledger state.

Manual smoke:

```sh
codewhale fleet run <sample-fleet-task> --max-workers 3
codewhale fleet status
codewhale fleet inspect <worker-id>
```

## Related

- #3167 Fleet profiles, slots, loadouts, permissions, and delegation.
- #3205 semantic route roles and Fleet model classes.
- #3384 atomic `ReadyRouteCandidate` route resolution.
- #3229 WhaleFlow coordination substrate and IR consumer.
- #3230 synthesis/reduce pass.
- #3166 Fleet smoke and soak test issue.
- #2981 v0.9 WhaleFlow umbrella.

**Comment by Hmbown** [comment]
Kimi Code audit note, 2026-06-12:

The fleet epic should keep the Kimi subagent lesson in view: scale comes from headless workers plus a strong control plane, not from making every child a heavy interactive mini-session.

Suggested framing for the fleet architecture:

- workers are headless `AgentWorkerSpec` executions by default;
- manager/scheduler/run-ledger own queueing, leases, heartbeats, status, receipts, and escalation;
- TUI/CLI/API are projections over the same worker events and run records;
- no invisible agent storm: launch pressure, queued counts, and blocked/rate-limited states must be visible before work runs;
- child richness belongs in profiles/capabilities/artifacts, not in UI-shaped session state.

This keeps Agent Fleet aligned with #3096 and gives Whaleflow `/swarm` a clean substrate instead of reintroducing the old `agent_swarm` failure mode.

**Comment by Hmbown** [comment]
Claude workflow signal from local release-harvest dogfood, 2026-06-12:

This epic should explicitly separate two related modes:

1. **Tactical workflow/swarm burst:** the parent agent discovers one bounded subproblem, starts multiple workers for that subproblem, and keeps working while they run.
2. **Agent Fleet:** an always-running control plane for durable queues, worker leases, alerts, remote hosts, and verifiable task specs.

The Claude example looked like mode 1:

- `Workflow(Map current TUI architecture for porting PR #3051 voice commands)`;
- background run, `0/6 agents done`, elapsed time/token usage visible;
- `/workflows` monitor/save path;
- parent agent immediately kicked off verification tests and task tracking rather than waiting.

That is not a replacement for Agent Fleet. It is the small, high-frequency product path that should share the fleet substrate. If CodeWhale gets this right, `/swarm`/workflow bursts become the daily-use affordance, while Fleet is the heavier durable operations layer.

Design note to add to the epic:

- Fleet protocol and ledger types should support both `tactical_workflow` and `fleet_run` scopes.
- Tactical workflows should inherit fleet primitives where useful: worker IDs, heartbeats, cancellation, receipts, token accounting, and inspect/save controls.
- Tactical workflows should not require users to pre-author a task spec or start a persistent manager; the parent agent can create them opportunistically from current context.

**Comment by Hmbown** [comment]
Retargeting the epic to v0.8.61. v0.8.60 now ships the local-first foundation: fleet protocol/ledger, scheduler/retry primitives, real `codewhale exec` worker cutover, CLI wrapper passthrough, docs tying sub-agents and fleet to one runtime, and release smoke evidence.

The full epic remains intentionally larger: SSH remote proof, long-lived manager/resume semantics, event-driven planner wakeups, day-scale runbooks, goal-controller integration, and richer API/TUI projections. Those should stay visible after v0.8.60 instead of blocking the release gate.

**Comment by Hmbown** [comment]
v0.8.61 has shipped on main (c9193e2e). The foundational parts of this issue are now in place; remaining work can be retargeted to v0.8.62. Updating milestone.

**Comment by Hmbown** [comment]
Pre-fix code map (2026-06-23)

Disposition: keep as the Fleet execution epic, but narrow it. The real-run substrate has largely shipped; this epic should now track proof/hardening around profiles, route tiers, worker runtime, and handoff receipts rather than asking agents to rebuild Fleet from scratch.

Likely remaining PR slices:
- Add `agent_profile` + route-tier/model-class schema (#3167/#3367/#3205).
- Ensure Fleet task specs carry profile/model-class intent into `AgentWorkerSpec` without prompt-only routing.
- Add real-run smoke/soak proof (#3166), including branch/PR handoff expectations.
- Harden ledger/event/status behavior for interrupted/failed/cancelled workers.

Code seams:
- `crates/protocol/src/fleet.rs:70` is the protocol task spec; `:109` is the current worker profile shape.
- `crates/tui/src/fleet/task_spec.rs:87` loads and validates TOML/JSON specs.
- `crates/tui/src/fleet/worker_runtime.rs:48` maps Fleet tasks to sub-agent worker specs.
- `crates/tui/src/fleet/worker_runtime.rs:154` maps model strings to `ModelRoute`; this should receive route tier/class intent once #3205 lands.
- `crates/tui/src/worker_profile.rs:126` is the runtime profile contract for permissions/tools/model/provider/background.

Acceptance checks:
- A Fleet worker runs on the intended branch/workspace with the intended profile and route tier.
- A worker can open a PR against `main` or the configured integration branch without losing Fleet run metadata.
- Sub-agent status/peek/cancel and ledger receipts agree for success, failure, interrupt, and budget exhaustion.
- No issue should ask for simulated completion where a real worker runtime path exists.

**Comment by Hmbown** [comment]
Resolved for v0.8.65: the profiled worker substrate landed across #3469, #3516, #3520, #3525, #3536, and the v0.8.65 integration branch.
